### PR TITLE
Add Conv2DBatchEnsemble layer and basic tests

### PR DESCRIPTION
**Pull request by runjerry**

### CODE REVIEW

File: alf/layers.py [hunk 1]
```diff
@@ -956,6 +956,161 @@ def bias(self):
...
+@alf.configurable
+class Conv2DBatchEnsemble(Conv2D):
+    r"""The BatchEnsemble for 2D Conv layer.
+
+    BatchEnsemble is proposed in `Wen et. al. BatchEnsemble: An Alternative Approach
+    to Efficient Ensemble and Lifelong Learning <https://arxiv.org/abs/2002.06715>`_
+
+    In a nutshell, a tuple of vector :math:`(r_k, s_k)` is maintained for ensemble
+    member k in addition to the original conv2d kernel W. For input x, the
+    result for ensemble member k is calculated as :math:`(W \circ (s_k r_k^T)) * x`.
```
Comment (emailweixu) [inline alf/layers.py]: For conv2d, W is rank-3 tensor. So this is not quite right.

**Review by emailweixu — CHANGES_REQUESTED** [review]

File: alf/layers.py (hunk 1 above)
Comment (runjerry) [reply alf/layers.py]: Corrected.

File: alf/layers.py [hunk 2]
```diff
@@ -956,6 +956,165 @@ def bias(self):
...
+                If a tuple, it should contain two tensors. The first one is the data 
+                tensor with shape ``[B, C, H, W]``. The second one is ensemble_ids 
+                indicating which ensemble member each sample should use. Its shape 
+                should be [batch_size], and all elements should be in [0, ensemble_size).
+        Returns:
+            tuple if ``output_ensemble_ids`` is True,
+            - Tensor: with shape as ``inputs.shape[:-1] + (output_size,)``
+            - LongTensor: if enseble_ids is provided, this is same as ``ensemble_ids``,
+                otherwise a randomly generated ensemble_ids is returned
+            Tensor if ``output_ensemble_ids`` is False. The result of FC.
```
Comment (emailweixu) [inline alf/layers.py]: The result of Conv2D

File: alf/layers.py [hunk 3]
```diff
@@ -956,6 +956,165 @@ def bias(self):
...
+        Args:
+            inputs (Tensor|tuple): if a Tensor, its shape should be ``[B, C, H, W]``.
+                And a random ensemble id will be generated for each sample in the batch. 
+                If a tuple, it should contain two tensors. The first one is the data 
+                tensor with shape ``[B, C, H, W]``. The second one is ensemble_ids 
+                indicating which ensemble member each sample should use. Its shape 
+                should be [batch_size], and all elements should be in [0, ensemble_size).
+        Returns:
+            tuple if ``output_ensemble_ids`` is True,
+            - Tensor: with shape as ``inputs.shape[:-1] + (output_size,)``
```
Comment (emailweixu) [inline alf/layers.py]: shape is not correct

File: alf/layers.py (hunk 2 above)
Comment (runjerry) [reply alf/layers.py]: Corrected.

File: alf/layers.py (hunk 3 above)
Comment (runjerry) [reply alf/layers.py]: Corrected.

**Review by emailweixu — APPROVED** [review]

**Pull request merged**